nx/lib/nx/lin_alg.ex

```diff
@@ -967,4 +967,59 @@ defmodule Nx.LinAlg do
       opts
     )
   end
+
+  @doc """
+  Produces the tensor taken to the given power by dot-product.
+
+  The input is always a square tensor and a non-negative integer,
+  and the output is a square tensor of the same dimensions as the input tensor.
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 6)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [5743, 8370],
+          [12555, 18298]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.eye(3), 65535)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1]
+        ]
+      >
+
+  """
+  @doc from_backend: false
+  def matrix_power(matrix, power) when is_integer(power) and power < 0 do
+    matrix_power(Nx.LinAlg.invert(matrix), -power)
+  end
+  def matrix_power(matrix, power) when is_integer(power) do
+    {result, _exp_matrix} = power
+    |> Integer.digits(2)
+    |> Enum.reverse()
+    |> Enum.reduce({Nx.eye(matrix), matrix}, fn
+      1, {result_matrix, exp_matrix} ->
+        {Nx.dot(result_matrix, exp_matrix), Nx.dot(exp_matrix, exp_matrix)}
```

Comment:
I did come to the realization that this algorithm potentially does one dot-product too many, namely the dot product to calculate the last (discarded) exponential matrix. This is also the most costly matrix to calculate, though again, with the maximum number width of 64 bits, this doesn't matter much.

Reply:
Suggestion above :)

Reply:
Also, you're probably missing a `mix format` in the code